pkg/engine/shield/add_test.go

```diff
@@ -0,0 +1,362 @@
+			},
+		},
+		{
+			ID: key.Shield("ShieldByTargetHP"),
+			Info: info.Shield{
+				Source:     source,
+				Target:     target,
+				BaseShield: info.ShieldMap{model.ShieldFormula_SHIELD_BY_TARGET_MAX_HP: 0.5},
+			},
+		},
+		{
+			ID: key.Shield("ShieldBySourceShield"),
+			Info: info.Shield{
+				Source:     source,
+				Target:     target,
+				BaseShield: info.ShieldMap{model.ShieldFormula_SHIELD_BY_SHIELDER_TOTAL_SHIELD: 0.5},
+			},
+		},
+	}
+
+	manager.event.ShieldAdded.Subscribe(func(event event.ShieldAdded) {
+		assert.Equal(t, 0.0, event.ShieldHealth)
+	})
+
+	for _, config := range shieldConfigs {
+		shieldID := config.ID
+		shieldInfo := config.Info
+		manager.AddShield(shieldID, shieldInfo)
+	}
+}
+
+func TestShieldHealthBy0Values(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	engine := mock.NewMockEngineWithEvents(mockCtrl)
+	attr := mock.NewMockAttribute(mockCtrl)
+	manager := New(engine.Events(), attr)
+
+	defer mockCtrl.Finish()
+
+	source := key.TargetID(1)
+	sourceStats := mock.NewEmptyStats(source)
+	sourceStats.AddProperty("tst", prop.ATKBase, 0.0)
+	sourceStats.AddProperty("tst", prop.DEFBase, 0.0)
+	sourceStats.AddProperty("tst", prop.HPBase, 0.0)
+	shield := &Instance{name: "SourceShield", hp: 0.0}
+	manager.targets[source] = append(manager.targets[source], shield)
+	attr.EXPECT().Stats(gomock.Eq(source)).Return(sourceStats).Times(5)
+
+	target := key.TargetID(2)
+	targetStats := mock.NewEmptyStats(target)
+	targetStats.AddProperty("tst", prop.HPBase, 0.0)
+	attr.EXPECT().Stats(gomock.Eq(target)).Return(targetStats).Times(5)
+
+	type shieldConfig struct {
+		ID   key.Shield
+		Info info.Shield
+	}
+
+	shieldConfigs := []shieldConfig{
+		{
+			ID: key.Shield("ShieldBySourceATK"),
+			Info: info.Shield{
+				Source:     source,
+				Target:     target,
+				BaseShield: info.ShieldMap{model.ShieldFormula_SHIELD_BY_SHIELDER_ATK: 0.5},
+			},
+		},
+		{
+			ID: key.Shield("ShieldBySourceDEF"),
+			Info: info.Shield{
+				Source:     source,
+				Target:     target,
+				BaseShield: info.ShieldMap{model.ShieldFormula_SHIELD_BY_SHIELDER_DEF: 0.5},
+			},
+		},
+		{
+			ID: key.Shield("ShieldBySourceHP"),
+			Info: info.Shield{
+				Source:     source,
+				Target:     target,
+				BaseShield: info.ShieldMap{model.ShieldFormula_SHIELD_BY_SHIELDER_MAX_HP: 0.5},
+			},
+		},
+		{
+			ID: key.Shield("ShieldByTargetHP"),
+			Info: info.Shield{
+				Source:     source,
+				Target:     target,
+				BaseShield: info.ShieldMap{model.ShieldFormula_SHIELD_BY_TARGET_MAX_HP: 0.5},
+			},
+		},
+		{
+			ID: key.Shield("ShieldBySourceShield"),
+			Info: info.Shield{
+				Source:     source,
+				Target:     target,
+				BaseShield: info.ShieldMap{model.ShieldFormula_SHIELD_BY_SHIELDER_TOTAL_SHIELD: 0.5},
+			},
+		},
+	}
+
+	manager.event.ShieldAdded.Subscribe(func(event event.ShieldAdded) {
+		assert.Fail(t, "A shield of 0 hp should not be added to the target, as such this event should never emit")
+	})
```

Comment:
This assert is failing (right now it is possible to add 0 HP shields and they do emit events)